### PR TITLE
[export] Add exportable attention and kv cache

### PR DESCRIPTION
**Pull request by larryliu0820**
Summary: As titled.

Test Plan: Rely on new unit tests

Reviewers:

Subscribers:

Tasks:

Tags:

#### Context
What is the purpose of this PR? Is it to
- [x] add a new feature
- [ ] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.

#### Changelog
What are the changes made in this PR?
* Adding `attention.py` in `torchtune/modules/_export/`
* Adding `kv_cache.py` in `torchtune/modules/_export/`
* Adding `test_attention.py` in `tests/torchtune/modules/_export/`

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [x] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [ ] run unit tests via `pytest tests`
- [ ] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [x] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

File: .github/workflows/unit_test.yaml [hunk 1]
```diff
@@ -34,7 +34,9 @@ jobs:
         run: python -m pip install --upgrade pip
       - name: Install dependencies
         run: |
-          python -m pip install torch torchvision torchao
+          NIGHTLY_VERSION=dev20241121
+          python -m pip install torch=="2.6.0.${NIGHTLY_VERSION}" torchvision=="0.20.0.${NIGHTLY_VERSION}" --extra-index-url https://download.pytorch.org/whl/nightly/cpu
+          python -m pip install torchao
```
Comment (ebsmothers) [inline .github/workflows/unit_test.yaml:39]: I don't think we want to change the default behavior of our unit test workflow here. Can we figure out a way to decouple the export CI a little more?

Comment (larryliu0820) [reply .github/workflows/unit_test.yaml:39]: Can you say a bit more? Are you ok with a separate CI job running these tests with torch nightly installed?

Comment (larryliu0820) [reply .github/workflows/unit_test.yaml]: Added `Export` CI workflow.

File: torchtune/modules/_export/install_requirements.sh [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+NIGHTLY_VERSION="dev20241121"
+
+# Install pytorch nightly for export-friendly modules to run.
+pip install torch==2.6.0.${NIGHTLY_VERSION} torchvision==0.20.0.${NIGHTLY_VERSION} --extra-index-url https://download.pytorch.org/whl/nightly/cpu
```
Comment (pbontrager) [inline torchtune/modules/_export/install_requirements.sh:11]: This sh file doesn't really make sense to me. The goal of the export folder is to house export friendly modules not modify the library setup. Currently torchtune operates under the assumption that the user installs pytorch themselves and some of our features our gated behind nightlies where we'll tell the user to upgrade to nightlies if they want to use that feature. The same pattern should be followed here.

File: tests/torchtune/modules/_export/test_attention.py [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+import unittest
+
+import torch
+from torch._inductor.package import load_package, package_aoti
+from torch.testing import assert_close
+from torchtune.models.llama3_1._position_embeddings import Llama3ScaledRoPE
+from torchtune.modules._export.attention import (
+    MultiHeadAttention as ExportMultiHeadAttention,
+)
+from torchtune.modules.attention import MultiHeadAttention as TTMultiHeadAttention
```
Comment (pbontrager) [inline tests/torchtune/modules/_export/test_attention.py:23]: This unit test should exist in a downstream library as per the original plan. Parity check tests (TTMultiHeadAttention == ExportMultiHeadAttention) can't exist in torchtune otherwise our PRs will be blocked whenever we make a change to MultiHeadAttention. Unit tests in torchtune (such as checking export) should be decoupled from implementations outside of the _export folder.

File: .github/workflows/export.yaml [hunk 4]
```diff
@@ -0,0 +1,43 @@
+name: Export
```
Comment (pbontrager) [inline .github/workflows/export.yaml:1]: We run tests on pytorch nightlies every day [flow](https://github.com/pytorch/torchtune/blob/main/.github/workflows/gpu_test.yaml). I'd rather not have a separate flow for export and have the tests run as part of the standard flows.

**Review by pbontrager — COMMENTED** [review]
Thanks for putting up this first commit. I'm not reviewing the modules themselves but the integration with unit tests as this is the first time we're doing this. I left some comments related to this.

File: .github/workflows/export.yaml (hunk 4 above)
Comment (larryliu0820) [reply .github/workflows/export.yaml:1]: So that was my first version, the only caveat is that I need to install nightly in the CI flow.

@ebsmothers commented on my first version:

> I don't think we want to change the default behavior of our unit test workflow here. Can we figure out a way to decouple the export CI a little more?

That's why I separated it out as a new CI flow.

File: .github/workflows/export.yaml [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  run:
+    shell: bash -l -eo pipefail {0}
+
+jobs:
+  export_unit_tests:
+    if: github.repository_owner == 'pytorch'
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        python-version: ['3.9', '3.10', '3.11']
```
Comment (dvorjackz) [inline .github/workflows/export.yaml:30]: ET doesn't explicitly support 3.9, should we be testing it?

File: tests/torchtune/modules/_export/test_attention.py [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+        assert_close(et_res, tt_res)
+
+        # test with kv cache
+        self.et_mha.setup_cache(1, dtype=torch.float32, max_seq_len=20)
+        self.tt_mha.setup_cache(1, dtype=torch.float32, max_seq_len=20)
+
+        et_res = self.et_mha(self.x, self.x)  # Self attention.
+        tt_res = self.tt_mha(self.x, self.x)  # Self attention.
+
+        self.assertTrue(torch.allclose(et_res, tt_res))
```
Comment (dvorjackz) [inline tests/torchtune/modules/_export/test_attention.py:123]:
```suggestion
        assert_close(et_res, tt_res)
```

File: tests/torchtune/modules/_export/test_attention.py [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+        self.tt_mha.reset_cache()
+
+        et_res = self.et_mha(
+            self.x, self.x, input_pos=self.input_pos
+        )  # Self attention with input pos.
+        tt_res = self.tt_mha(
+            self.x, self.x, input_pos=self.input_pos
+        )  # Self attention with input pos.
+
+        self.assertTrue(torch.allclose(et_res, tt_res))
```
Comment (dvorjackz) [inline tests/torchtune/modules/_export/test_attention.py:134]:
```suggestion
        assert_close(et_res, tt_res)
```

File: tests/torchtune/modules/_export/test_attention.py [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+        mask = self.causal_mask[self.input_pos, :]
+        # First run
+        et_res = self.et_mha(
+            self.x, self.x, mask=mask, input_pos=self.input_pos
+        )  # Self attention with input pos.
+        tt_res = self.tt_mha(
+            self.x, self.x, mask=mask, input_pos=self.input_pos
+        )  # Self attention with input pos.
+
+        self.assertTrue(torch.allclose(et_res, tt_res))
```
Comment (dvorjackz) [inline tests/torchtune/modules/_export/test_attention.py:216]:
```suggestion
       assert_close(et_res, tt_res)
```

**Review by dvorjackz — COMMENTED** [review]
Should we also copy over the [README](https://github.com/pytorch/executorch/blob/main/extension/llm/modules/README.md)?

File: .github/workflows/export.yaml (hunk 5 above)
Comment (larryliu0820) [reply .github/workflows/export.yaml:30]: This is testing export not ET

**Review by ebsmothers — APPROVED** [review]
